### PR TITLE
Add the new structure role optional option to the gadget spec schema.

### PR DESCRIPTION
**Pull request by sil2100**
LP: #1638660

### CODE REVIEW

**Comment by sil2100** [comment]
Is there anything else we need to do with the role attribute? From the gadget spec I only saw it adding this one additional size constraint.

**Comment by warsaw** [comment]
@sil2100 Thanks for the branch!  I'll review the code in a bit, but yes there is another thing we have to do for this.  In the old spec, a structure's `type` could take a literal string "mbr".  This functionality is subsumed by `role: mbr`, so we need to change the code that checks for `type == 'mbr'` to now check for the role, and also remove the parsing of the special `type: mbr` part.

We also need to handle `role: system-boot` and `role: system-data`.  For example, if you look in builder.py, we match on `filesystem_label == 'system-boot'` to identify the boot partition.  This needs to change to check for `role: system-boot`.

We also hard code the location of the system-data (see `populate_rootfs_contents()`).  This of course needs to check for `role: system-data`.

It's a fairly big change over, so if you just want to handle the parsing bits, that's cool.  We can make the changes into two branches.

File: ubuntu_image/parser.py [hunk 1]
```diff
@@ -51,6 +51,12 @@ class FileSystemType(Enum):
     vfat = 'vfat'
 
 
+class StructureRole(Enum):
+    mbr = 'mbr'
+    systemboot = 'system-boot'
+    systemdata = 'system-data'
```
Comment (warsaw) [inline ubuntu_image/parser.py]: I think the attribute names should be `system_boot` and `system_data`.  I know the other example does use `uboot` instead of `u_boot` but that's okay because "u" isn't itself a word.

**Comment by sil2100** [comment]
Ok, changed the StructureRole fields as per comment, will also modify the parser code as per instructions in a moment. We can do it in two turns - but, if possible, I would also like to maybe try and work on the rest of the changes. This would give me some additional experience in the other parts of the code beyond the parser :)

**Comment by sil2100** [comment]
Ok, pushed the changes to the parser code as mentioned. I guess I'll just use this branch as a work-place for getting everything changed - but the parser parts should be good for review now.

**Comment by warsaw** [comment]
Sorry to do that to you, but could you rebase and fix the conflicts?

**Comment by warsaw** [comment]
I see that the tests are failing.  I need to check if the actual gadget.yamls have caught up to the new specification.  I expressed my concern about a migration path, and we'll soon introduce a format version tag, but for now we might have to be more loosy-goosy in our specification parsing.  I will test this locally and get back to you.

**Comment by warsaw** [comment]
You need this change in order to sbuild the package, since some of the new tests require network access (which isn't available in sbuild).

```
1 file changed, 7 insertions(+), 1 deletion(-)
ubuntu_image/tests/test_main.py | 8 +++++++-

modified   ubuntu_image/tests/test_main.py
@@ -254,6 +254,8 @@ class TestMainWithModel(TestCase):
 
 
 class TestMainWithBadGadget(TestCase):
+    maxDiff = None
+
     def setUp(self):
         super().setUp()
         self._resources = ExitStack()
@@ -261,6 +263,8 @@ class TestMainWithBadGadget(TestCase):
         self.model_assertion = resource_filename(
             'ubuntu_image.tests.data', 'model.assertion')
 
+    @skipIf('UBUNTU_IMAGE_TESTS_NO_NETWORK' in os.environ,
+            'Cannot run this test without network access')
     def test_bad_gadget_log(self):
         log = self._resources.enter_context(LogCapture())
         workdir = self._resources.enter_context(TemporaryDirectory())
@@ -276,7 +280,9 @@ class TestMainWithBadGadget(TestCase):
             (logging.ERROR, 'Use --debug for more information')
             ])
 
-    def test_bad_gadget_debug_log(self):
+    @skipIf('UBUNTU_IMAGE_TESTS_NO_NETWORK' in os.environ,
+            'Cannot run this test without network access')
+    def test_bad_gadget_log_debug(self):
         log = self._resources.enter_context(LogCapture())
         workdir = self._resources.enter_context(TemporaryDirectory())
         self._resources.enter_context(patch(
```

**Comment by warsaw** [comment]
This looks pretty good.  I want to do a little bit more review of the code, but there's another problem we have to deal with.  The official gadgets haven't been updated to this new schema yet, so the autopkgtests fail.

```
gadget.yaml parse error: Invalid gadget.yaml @ volumes:pc:structure:0:type
Use --debug for more information
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/ubuntu_image/parser.py", line 292, in parse
    validated = GadgetYAML(yaml)
  File "/usr/lib/python3/dist-packages/voluptuous/schema_builder.py", line 192, in __call__
    return self._compiled([], data)
  File "/usr/lib/python3/dist-packages/voluptuous/schema_builder.py", line 486, in validate_dict
    return base_validate(path, iteritems(data), out)
  File "/usr/lib/python3/dist-packages/voluptuous/schema_builder.py", line 324, in validate_mapping
    raise er.MultipleInvalid(errors)
voluptuous.error.MultipleInvalid: expected HybridId for dictionary value @ data['volumes']['pc']['structure'][0]['type']

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/autopkgtest.WgcN8Y/build.AoF/ubuntu-image-0.12+17.04ubuntu1/debian/tests/mount", line 44, in <module>
    gadget = parse(fp.read())
  File "/usr/lib/python3/dist-packages/ubuntu_image/parser.py", line 300, in parse
    raise GadgetSpecificationError('Invalid gadget.yaml @ {}'.format(path))
ubuntu_image.parser.GadgetSpecificationError: Invalid gadget.yaml @ volumes:pc:structure:0:type
```

(I guess I should figure out why that's printing the full traceback when it should just print the error.)

In any case, this probably means we have to continue to accept but deprecate `type: mbr`.  We could potentially issue a `DeprecationWarning` when we see that.

**Comment by warsaw** [comment]
So yes, the gadget.yaml from the pc-amd64-model.assertion still contains a `type: mbr`.  Looking back over my notes with @niemeyer we will have to accept that and issue a `DeprecationWarning` for now.  Eventually we'll reject such gadget.yaml files.

However, we should not accept a gadget.yaml that contains both `type: mbr` and `role: mbr`.

We can defer handling of `role: system-data` and `role: system-boot` for a follow up branch.

https://github.com/snapcore/snapd/wiki/Gadget-snap#gadget.yaml

File: ubuntu_image/tests/test_parser.py [hunk 2]
```diff
@@ -308,37 +309,7 @@ def test_size_offset_suffix(self):
         partition0 = volume0.structures[0]
         self.assertEqual(partition0.size, MiB(3))
 
-    def test_mbr_structure(self):
```
Comment (warsaw) [inline ubuntu_image/tests/test_parser.py]: We'll probably want to keep these tests of course, to verify legacy (deprecated) behavior.

File: ubuntu_image/tests/test_parser.py [hunk 3]
```diff
@@ -347,12 +318,11 @@ def test_mbr_structure_conflicting_filesystem(self):
...
-        - type: mbr
-          size: 400M
-          filesystem: ext4
+        - type: 00000000-0000-0000-0000-0000deadbeef
 """)
-        self.assertEqual(str(cm.exception),
-                         'mbr type must not specify a file system')
+        self.assertEqual(
+            str(cm.exception),
+            'Invalid gadget.yaml @ volumes:first-image:structure:0:size')
```
Comment (warsaw) [inline ubuntu_image/tests/test_parser.py]: This seems wrong, since the test is specifically checking that `type: mbr` is incompatible with `filesystem:`.  That should probably just change to `role: mbr` and continue to verify the incompatibility with `filesystem:`.

Although, hmm, it looks like you might have added such a test below.  In which case, because of the legacy behavior, just keep this one unchanged.

File: ubuntu_image/tests/test_parser.py [hunk 4]
```diff
@@ -526,6 +496,151 @@ def test_volume_filesystem_default_none(self):
...
+            cm = resources.enter_context(
+                self.assertRaises(GadgetSpecificationError))
+            parse("""\
+volumes:
+  first-image:
+    schema: mbr
+    bootloader: u-boot
+    structure:
+        - type: ef
+          size: 1M
```
Comment (warsaw) [inline ubuntu_image/tests/test_parser.py]: Can you add some corner case tests, e.g. with size == (445, 446, 447) where the first two are allowed, but the last one is not?

File: ubuntu_image/tests/test_parser.py (hunk 2 above)
Comment (sil2100) [inline ubuntu_image/tests/test_parser.py]: I removed this test as I added earlier the test_volume_structure_role that already tested the role: mbr part - but yeah, since now we need to support both the new and the old, I'll re-introduce this test but maybe with a different name.

**Comment by sil2100** [comment]
ACK on all the changes! Yeah, will clean it up now that I know that the old code-paths still need to be considered. Thanks for the review!

**Comment by warsaw** [comment]
I'm going to push the `skipIf()` changes to master, since we need them there anyway.

**Comment by warsaw** [comment]
@sil2100 I've landed a slew of build/test improvements to master.  Please rebase and we'll see if that fixes your checks!  I think we can move quickly from there.  Thanks again for the great work.

**Comment by sil2100** [comment]
Hey! First of all: apologies for these proposed changes, I'll understand if you disagree with them and correct as needed

 This is my proposition of handling the type: mbr special case - I assume that for the mbr type we have the same constraints as for the mbr role, so I just print out a deprecation warning and assume it's role = mbr in such a case. Of course I might be wrong, but I didn't want to unnecessarily add type == 'mbr' conditionals in addition to the already role-based ones.

Also, ekhm, I wanted to write a nice test-case for the edge-cases in the mbr size test. As I'm not an expert in these things, such a test can probably be written in a better way - I just didn't want to boilerplate 3 test-cases with the same thing.

**Comment by sil2100** [comment]
Ok, by the failing mount test I see that the role: mbr restrictions were not upheld for cases where type: mbr was used, for instance it seems to have been allowed to have a non-mbr volume schema with an mbr partition type? Seems like I'll have to just introduce handling type == 'mbr' separately. So long nicely looking code.

**Comment by warsaw** [comment]
For example, here's the gadget.yaml from the amd64 model that the mount test uses:

```
volumes:
  pc:
    bootloader: grub
    structure:
      - name: mbr
        type: mbr
        size: 440
        content:
          - image: pc-boot.img
      - name: BIOS Boot
        type: DA,21686148-6449-6E6F-744E-656564454649
        size: 1M
        offset: 1M
        offset-write: mbr+92
        content:
          - image: pc-core.img
      - name: EFI System
        type: EF,C12A7328-F81F-11D2-BA4B-00A0C93EC93B
        filesystem: vfat
        filesystem-label: system-boot
        size: 50M
        content:
          - source: grubx64.efi
            target: EFI/boot/grubx64.efi
          - source: shim.efi.signed
            target: EFI/boot/bootx64.efi
          - source: grub.cfg
            target: EFI/ubuntu/grub.cfg
```

Clearly there's no volume schema, so that defaults to gpt.   And the first structure has a type of `mbr`.  I'd say this is a bug in the gadget.yaml, but since we used to support it, I suppose for legacy mode we'll have to continue to support it.  That's unfortunately because I liked your suggestion, although even treating type=mbr the same as role=mbr causes a minor problem that the error reporting was wrong.

I'd like to get @vorlonofportland 's opinion on this too.

File: ubuntu_image/parser.py [hunk 5]
```diff
@@ -301,41 +313,59 @@ def parse(stream_or_string):
...
+            # happens above in the Voluptuous schema, but here we need to
+            # ensure cross-attribute constraints.  Specifically, hybrids are
+            # allowed for either schema, but GUID-only is only allowed for GPT,
+            # while 2-digit-only is only allowed for MBR roles.
+            # Note too that 2-item tuples are also already ensured.
+            # We also still support the deprecated special type 'mbr' for
+            # legacy purposes, but issue a warning
+            if structure_type == 'mbr':
+                warn('Structure type mbr is deprecated, please use structure '
+                     'roles instead', DeprecationWarning)
```
Comment (warsaw) [inline ubuntu_image/parser.py]: What about using the same nomenclature we use for GadgetSpecificationErrors?  E.g.

```
warn('volumes:<volume name>:structure:<N>:type = 'mbr' is deprecated; use role instead')
```

Or does that look too ugly?

File: ubuntu_image/parser.py [hunk 6]
```diff
@@ -320,41 +332,62 @@ def parse(stream_or_string):
...
             structure_type = structure['type']
-            # Validate structure types.  These can be either GUIDs, two hex
-            # digits, hybrids, or the special 'mbr' type.  The basic syntactic
-            # validation happens above in the Voluptuous schema, but here we
-            # need to ensure cross-attribute constraints.  Specifically,
-            # hybrids and 'mbr' are allowed for either schema, but GUID-only
-            # is only allowed for GPT, while 2-digit-only is only allowed for
-            # MBR.  Note too that 2-item tuples are also already ensured.
+            structure_role = structure.get('role')
+            is_mbr = \
```
Comment (warsaw) [inline ubuntu_image/parser.py]: Preferably no backslashes.  I think you can write this as:

```
    is_mbr = (structure_type == 'mbr'
                    or structure_role is StructureRole.mbr)
```

modulo proper formatting that GH's review window apparently doesn't support ;)

**Comment by vorlonofportland** [comment]
As discussed via email: schema: mbr and role: mbr (previously: type: mbr) must not be conflated.  It is perfectly valid to use schema: gpt (which tells what type of partition table to use) together with role: mbr (which says to write some data to the boot sector of the disk).

File: ubuntu_image/parser.py [hunk 7]
```diff
@@ -320,41 +332,62 @@ def parse(stream_or_string):
...
+            # Note too that 2-item tuples are also already ensured.
+            # We also still support the deprecated special type 'mbr' for
+            # legacy purposes, but issue a warning
+            if structure_type == 'mbr':
+                warn("volumes:<volume name>:structure:<N>:type = 'mbr' is "
+                     'deprecated; use role instead', DeprecationWarning)
+                if structure_role:
+                    raise GadgetSpecificationError(
+                        'Type mbr and role fields assigned at the same time, '
+                        'please use role field only')
```
Comment (vorlonofportland) [inline ubuntu_image/parser.py]: Imprecise error message, you don't check that structure_role is set to mbr here so 'please use role field only' could be directing them to do something that is syntactically correct but semantically not at all what they meant.

File: ubuntu_image/parser.py [hunk 8]
```diff
@@ -320,41 +332,62 @@ def parse(stream_or_string):
...
                     offset = MiB(1)
                 else:
                     offset = last_offset
             last_offset = offset + size
             # Extract the rest of the structure data.
             structure_id = structure.get('id')
             filesystem = structure['filesystem']
-            if structure_type == 'mbr':
-                if structure_id is not None:
+            if structure_role is StructureRole.mbr:
```
Comment (vorlonofportland) [inline ubuntu_image/parser.py]: Some places you check structure_role, some places you check structure_role and structure_type, some places you use is_mbr.  I think that after the deprecation checks, you should just promote type == mbr to structure_role, and check only structure_role from then on; then we don't have to rewrite this code when obsoleting the old syntax.

File: ubuntu_image/parser.py [hunk 9]
```diff
@@ -320,41 +332,62 @@ def parse(stream_or_string):
...
+                if size > 446:
                     raise GadgetSpecificationError(
-                        'mbr type must not specify partition id')
-                elif filesystem is not FileSystemType.none:
+                        'mbr role structures cannot be larger than 446 bytes.')
+                if schema is not VolumeSchema.mbr:
```
Comment (vorlonofportland) [inline ubuntu_image/parser.py]: as discussed, this is incorrect.

File: ubuntu_image/parser.py (hunk 8 above)
Comment (sil2100) [inline ubuntu_image/parser.py]: Yes, if you see previous commits this is how it was done originally, this was the 'nice' solution I did. This has been changed as we thought that there are differences between how the legacy 'mbr' type was handled and the new mbr role - which, as per the mailing thread, has been resolved.

File: ubuntu_image/parser.py [hunk 10]
```diff
@@ -320,41 +332,56 @@ def parse(stream_or_string):
...
+            # ensure cross-attribute constraints.  Specifically, hybrids are
+            # allowed for either schema, but GUID-only is only allowed for GPT,
+            # while 2-digit-only is only allowed for MBR roles.
+            # Note too that 2-item tuples are also already ensured.
+            # We also still support the deprecated special type 'mbr' for
+            # legacy purposes, but issue a warning
+            if structure_type == 'mbr':
+                warn("volumes:<volume name>:structure:<N>:type = 'mbr' is "
+                     'deprecated; use role instead', DeprecationWarning)
+                if structure_role:
```
Comment (warsaw) [inline ubuntu_image/parser.py]: Maybe we should move this to above the warning?  The way it is now, if you define both role and type, you'll get the deprecation warning and then the error.  I think it would be better to just show the error (since they're both encouraging the user to switch to role).

File: ubuntu_image/parser.py [hunk 11]
```diff
@@ -320,41 +332,56 @@ def parse(stream_or_string):
...
+                    raise GadgetSpecificationError(
+                        'Type mbr and role fields assigned at the same time, '
+                        'please use the mbr role instead')
+                structure_role = StructureRole.mbr
             if (isinstance(structure_type, UUID) and
                     schema is not VolumeSchema.gpt):
                 raise GadgetSpecificationError(
                     'MBR structure type with non-MBR schema')
             elif (isinstance(structure_type, str) and
```
Comment (warsaw) [inline ubuntu_image/parser.py]: It will be nice when we can get rid of `type='mbr'` so that the logic here can be simplified.  I had to work out the cases in my head again to convince myself that these checks are still correct, and I think they are.  I'd probably add a comment above this to explain (or remind us ;) what's going on, but don't worry about that.  I will do this after the branch lands.

**Comment by warsaw** [comment]
@vorlonofportland thanks for weighing in.  I have just one small change request and then I'm good with landing this.  I'll want you to approve the changes too before I accept the merge request.

**Comment by warsaw** [comment]
@vorlonofportland @sil2100 We've discussed whether it makes more sense to squash merge this branch.  I'm inclined to agree, just to keep a cleaner history, and just this once.  I'd have to re-enable it for this repo and then turn it back off once this branch lands.  Unless @vorlonofportland has a strong opinion otherwise, I will do this once both of our reviews are approved and the tests pass.

**Comment by warsaw** [comment]
Rather than put @sil2100 through rebase hell, I'll just merge this.  Clean history be damned.
